### ISSUE TITLE
`connmanctl` phantom input if interrupted with ctrl-c

### ISSUE THREAD

**Issue by kwshi** [elves/elvish#706]
If `connmanctl` is started, then killed with Ctrl-C, then started again, input keystrokes are not rendered to the screen---though still somehow registered and processed, *a la* password input.

More precisely, here's what happens:

- `connmanctl`: behaves normally, registering input as expected
- ctrl-c to interrupt
- `connmanctl` again, and typing some keys: none of the keys are rendered to the screen, but they are registered "internally" correctly since pressing return sends an actual command and causes connmanctl to respond with the corresponding output.

This behavior is not exhibited when `connmanctl` is launched from bash.

Perhaps there is something off about how connmanctl handles stdin, and that elvish is not prepared for it?  Other interactive tools that capture stdin, such as `cat` (I can't think of anything else to test with, for now) also do not exhibit this behavior, suggesting that the issue is particular to connmanctl, or whatever it uses to handle stdin.

**Comment by xiaq** [comment]
Connmanctl seems to be an Archlinux-specific tool so I cannot reproduce this myself. Can you give me the following information:

* After the bug is triggered, does other commands like `cat` behave correctly?
* What is the output of `stty` on a "clean" terminal versus after the bug is triggerd?

My current suspicion is that connmanctl left the terminal in a weird state.

**Comment by kwshi** [comment]
cat works correctly. I can't think of other stdin-ish tools to test this behavior with.  I'll try the `stty` as soon as I get back to my laptop.  

The error does seem particular to connmanctl, so it's probably true that connmanctl particular does something weird to the terminal.  It seems that other shells seem able to cope with this weirdness, though.

P.S. What linux do you use?

**Comment by occivink** [comment]
@xiaq I don't think `connmanctl` is archlinux-specific, it's a CLI for the `connman` connection manager.

**Comment by kwshi** [comment]
@xiaq Okay, here is what happens with `stty`.

In Elvish:
```
~> stty
speed 38400 baud; line = 0;
-brkint -imaxbel iutf8
~> connmanctl
connmanctl> Exception: connmanctl killed by signal interrupt
[tty], line 1: connmanctl
~> stty
speed 38400 baud; line = 0;
lnext = <undef>; min = 1; time = 0;
-brkint -icrnl -imaxbel iutf8
-icanon -echo
```

In Bash:
```
[kshi@dexi ~]$ stty
speed 38400 baud; line = 0;
-brkint -imaxbel iutf8
[kshi@dexi ~]$ connmanctl
connmanctl> 
[kshi@dexi ~]$ stty
speed 38400 baud; line = 0;
-brkint -imaxbel iutf8
```

Ah, so I've noticed that various other utilities that rely on stdin also break as soon as the connmanctl-interrupt "pollutes" the terminal (they themselves don't seem to cause the polluting, though it's a little hard to tell because everything I've tried, i.e. python, bash, ipython, catch the interrupt themselves).

**Comment by xiaq** [comment]
OK, this now confirms my suspicion that `connmanctl` is putting the terminal in non-canonical mode and it does not restore the terminal when being interrupted. This seems to be a bug of `connmanctl` but Elvish can do something to mitigate such situations.

Right now, Elvish saves the terminal attributes every time it activates the line editor and restores the attributes when it deactivates the line editor (e.g. to run a command):

https://github.com/elves/elvish/blob/14feeb8c1b3b880b84f382119148faa774c32555/edit/edcore/edit.go#L309

https://github.com/elves/elvish/blob/14feeb8c1b3b880b84f382119148faa774c32555/edit/edcore/edit.go#L349

This means that when `connmanctl` is interrupted, leaving the terminal in non-canonical mode, Elvish remembers that state, and puts the terminal back in non-canonical mode when running the next command.

We can change this logic to remember the terminal state only once when Elvish starts, and reuse that terminal state every time.

**Comment by xiaq** [comment]
This should be fixed now. @kwshi, reopen if you can still reproduce this.

**Comment by krader1961** [comment]
Note that Elvish has to assume the tty is in a weird state after resuming control from any external command. Even a well behaved command that restored the tty state might be unable to do so because it died from calling `abort()` or receiving `SIGSEGV` or `SIGKILL`. Also keep in mind that the user might want to use the `stty` command to change tty attributes such as the interrupt key. That is pretty rare these days but not unheard of. See, for example, this [fish shell issue](https://github.com/fish-shell/fish-shell/issues/2315).

**Comment by xiaq** [comment]
@krader1961 Good point about having well-behaved command leaving the terminal in a bad state.

For the second point, do you mean that Elvish should still make it possible for the user to use `stty` to control terminal attributes?

It seems to me that 1) guarding against ill-behaving command messing up with the terminal and 2) supporting use of `stty` are two conflict goals. I don't think you can really distinguish commands messing with the terminal by accident and commands messing with the terminal on purpose. We either respect what the command does to the terminal, and we lose 1; or we ignore what the command does to the terminal and always restores the terminal to a state we know is good, and we lose 2.

There may be a middle ground, namely respecting *parts* of the terminal setting external commands leaves and restore the others. But that would be pretty inconsistent - for instance, the user might wonder why `stty flagx` works but `stty flagb` does not.

Another interesting option is to provide a command that explicitly signals "run this command, and retain the terminal settings it make" to make `stty` still work to some extent.

**Comment by krader1961** [comment]
The fish shell issue I linked to has two main schools of thought. I have seen people propose that fish should assume the tty is in a sane state and if it isn't require the user to type things like `stty sane [ctrl-j]`. Even for a grey beard like myself who was accustomed to typing that at least once a week back in the 1990's that proposal is absurd. The more sensible approaches are:

1) Use a whitelist of tty attributes the user can change. Things like the magic chars like `stty intr` (the `c_cc[VINTR]` termios structure member) would be whitelisted.

2) Use a blacklist of tty attributes it never makes sense for the user to change. This would include things like the `ICANON` flag.

Option 2 gives the user maximal control but also makes it more likely that a program that fails to exit cleanly or a `stty` command with a stupid set of options will leave the tty in a minimally usable but weird state. I prefer option 1. Primarily because in practice the only things these days that users seem to want to change are the canonical mode special chars. I can't recall the last time I needed to, or heard of anyone else needing to, change attributes like the parity, or enable translating uppercase to lowercase, or having the tty driver convert tabs to spaces. The only exception to this is enabling or disabling software flow control.

The problem with a way to signal to elvish that it should "run this command, and retain the terminal settings it makes" is it will be hard to discover. Yes, to the extent the whitelist and blacklist approaches might surprise a user who deliberately attempts to do `stty -icanon`, for example, I would argue that user deserves to be surprised :smile: The solution might be to display a warning message that the tty was not in a sane state and elvish corrected the problem.
